helix-term/src/handlers/diagnostics.rs

```diff
@@ -21,4 +32,201 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
         }
         Ok(())
     });
+
+    let tx = handlers.pull_diagnostics.clone();
+    register_hook!(move |event: &mut DocumentDidChange<'_>| {
+        if event
+            .doc
+            .has_language_server_with_feature(LanguageServerFeature::PullDiagnostics)
+        {
+            let document_id = event.doc.id();
+            send_blocking(&tx, PullDiagnosticsEvent { document_id });
+        }
+        Ok(())
+    });
+
+    register_hook!(move |event: &mut DocumentDidOpen<'_>| {
+        if event
+            .doc
+            .has_language_server_with_feature(LanguageServerFeature::PullDiagnostics)
+        {
+            let document_id = event.doc.id();
+            job::dispatch_blocking(move |editor, _| {
+                let Some(doc) = editor.document(document_id) else {
+                    return;
+                };
+
+                let language_servers =
+                    doc.language_servers_with_feature(LanguageServerFeature::PullDiagnostics);
+
+                for language_server in language_servers {
+                    pull_diagnostics_for_document(doc, language_server);
+                }
+            })
+        }
+
+        Ok(())
+    });
+}
+
+#[derive(Debug)]
+pub(super) struct PullDiagnosticsHandler {
+    document_id: Option<DocumentId>,
```

Comment:
It should be possible for multiple documents to change while the changes are debounced. (For example the language server might send edits for multiple documents for an LSP rename request.) Instead of tracking one document I believe we need a HashSet/BTreeSet of document IDs to request

Reply:
I did not think of that scenario. 
I chose to use `HashSet`. The dataset will be very small and short lived so i am not sure what would be the best.